### DISCUSSION TITLE
kb/software/dijkstra

### DISCUSSION THREAD

**Discussion by giscus** [jiegec/kb#26] [General]
# kb/software/dijkstra

Knowledge Base of @jiegec

https://jia.je/kb/software/dijkstra.html

**Comment by bobofabtju** [comment]
您好，看到您写的文章《https://jia.je/hardware/2024/07/30/debian-linux-on-surface-laptop-7/》，我手头也有一台laptop，按照您的文章操作了一下，还是一直有问题，主要的问题如下：
1. 如果更新initrd文件中的内核deb文件和modules，生成的文件会超过installer.img的300M的限制；
2. 只更新部分文件仍然无法识别键盘；
所以想请问一下，是否方便分享一下您安装的img文件，非常感谢

**Reply by jiegec** [reply]
1. 我删掉了一些不需要的东西，避开了这个容量限制。
2. 键盘需要打一些 patch，并且用对应的 dts，见 https://github.com/jiegec/linux/commits/surface-laptop-7/。在打上 patch 之前，建议用 USB 键盘。

**Reply by bobofabtju** [reply]
感谢回复！
我修改了platform和kernel debug，目前也避开了容量限制，type-c键盘也可以识别了
但是disk partition搞定后，卡在了install base system这一步，提示”The base system installation into /target/ Failed“，不知道您是否有遇到过这个问题，非常感谢

**Reply by bobofabtju** [reply]
我目前img的主要内容基于debian-12-installer.img，主要改动如下：
1. 基于linux-6.11-rc4内核，编译生成了dtb，vmlinuz(Image.gz), linux_image_xxx.deb文件和modules目录；
2. 在initrd中，更新了dtb、linux_image_xxx.deb文件，并更新了lib/modules/6.11-rc4(删除了其他版本的module)，并更改了/usr/lib/finish-install.d/60install-custom-kernel
中的deb和dtb文件名；利用cpio和gzip压缩生成initrd.gz，并更新至img文件中；
3. 更新img目录中的vmlinuz，dtb文件，并修改grub.cfg中的command
不知道是否缺少了了哪个步骤，请指导一下，非常感谢

**Reply by jiegec** [reply]
> 感谢回复！ 我修改了platform和kernel debug，目前也避开了容量限制，type-c键盘也可以识别了 但是disk partition搞定后，卡在了install base system这一步，提示”The base system installation into /target/ Failed“，不知道您是否有遇到过这个问题，非常感谢

不太记得了，不过你可以切换 tty 去看 debian installer 的日志。这个版本的 debian installer 比较老，又因为是 testing，可能会有一些 bug，我最后是切换到 root shell 的 tty 再 chroot 进去手动完成剩下的安装步骤。其实或许改造其他发行版的 installer image 会更简单，例如换掉 archlinuxarm 的内核，然后手动安装。

**Reply by bobofabtju** [reply]
非常感谢回复
手动安装对我来说有一些陌生，我搜索了一个教程，请问可以参考这个流程吗？
https://developer.aliyun.com/article/1194974

**Reply by bobofabtju** [reply]
不知道是否方便添加您的联系方式，曾经给您介绍中的邮箱发送过邮件，但是没有成功；
所以如果方便的话，麻烦添加我的微信：18810514597，非常感谢

**Reply by jiegec** [reply]
> 非常感谢回复 手动安装对我来说有一些陌生，我搜索了一个教程，请问可以参考这个流程吗？ https://developer.aliyun.com/article/1194974

可以按照这个做，注意架构是 arm64，最后要替换掉内核+dts+grub配置，其他没啥了。

**Reply by bobofabtju** [reply]
抱歉又来打扰您，我按照手动教程在chroot中完成了所有的步骤，也将boot目录中的vmlinu+deb+grub文件夹都替换了
但是现在重启后会自动进入windows，无法进入linux，请问我需要重新生成grub的binary吗？
感觉我对grub的处理还有些问题

**Reply by jiegec** [reply]
> 抱歉又来打扰您，我按照手动教程在chroot中完成了所有的步骤，也将boot目录中的vmlinu+deb+grub文件夹都替换了 但是现在重启后会自动进入windows，无法进入linux，请问我需要重新生成grub的binary吗？ 感觉我对grub的处理还有些问题

需要 grub-install 在 UEFI 中添加启动项

**Reply by bobofabtju** [reply]
始终还是搞不定，按照教程中执行grub会一直报错。。。。

**Reply by bobofabtju** [reply]
非常抱歉，有两个问题要请教一下：
1. 我现在是把laptop自带的硬盘分出来100g，原本的windows的efi是在/dev/nvme0n1p1，debian的boot目录是挂在/dev/nvme0n1p4，如果想要在uefi中添加启动项，是应该在windows中操作，还是在debian中操作？我是不是有必要把两个EFI的目录放到一起？
2. 在debian中运行grub时，我看您说要替换grub配置，我是按照常规流程走完，替换掉/boot/grub中的cfg文件，还是直接把installer中的grub目录cpy过来？

**Reply by jiegec** [reply]
> 非常抱歉，有两个问题要请教一下：
> 
>     1. 我现在是把laptop自带的硬盘分出来100g，原本的windows的efi是在/dev/nvme0n1p1，debian的boot目录是挂在/dev/nvme0n1p4，如果想要在uefi中添加启动项，是应该在windows中操作，还是在debian中操作？我是不是有必要把两个EFI的目录放到一起？
> 

Debian。没有必要，放一个分区内的不同目录，不同分区，都可以，你可以在 linux 下用 efibootmgr 查看启动项。

>     2. 在debian中运行grub时，我看您说要替换grub配置，我是按照常规流程走完，替换掉/boot/grub中的cfg文件，还是直接把installer中的grub目录cpy过来？

你可以参考 linaro 的在那个 image 下的文档，修改 grub 的生成配置，让他自动往配置里写入 devicetree 命令，剩下就是 grub-mkconfig 自动生成配置。你也可以手动改，只要确认 kernel，cmdline，devicetree 都是对的就可以。

**Reply by bobofabtju** [reply]
哈喽，感谢之前的指导，我目前通过改u盘的grub set已经可以启动到debian，但是执行grub-install的时候还是会失败；会提示can not read efi boot* variables
导致不插u盘就无法引导到debian的grub，请问您遇到过这个问题吗？

**Reply by jiegec** [reply]
> 哈喽，感谢之前的指导，我目前通过改u盘的grub set已经可以启动到debian，但是执行grub-install的时候还是会失败；会提示can not read efi boot* variables 导致不插u盘就无法引导到debian的grub，请问您遇到过这个问题吗？

检查一下 efivarfs 是否挂载

**Reply by bobofabtju** [reply]
非常感谢耐心指导，已经可以稳定启动debian了，有个问题，您那边可以使用wifi吗？可以安装desktop吗？

**Reply by jiegec** [reply]
> 非常感谢耐心指导，已经可以稳定启动debian了，有个问题，您那边可以使用wifi吗？可以安装desktop吗？

wifi 时好时坏，建议先走 usb 以太网。desktop 没问题，就是显卡加速还要等等，但 cpu 性能也足够，不会卡顿。
